File: src/test/java/org/springframework/samples/petclinic/e2e/pages/OwnerFormPage.java
/*
 * Copyright 2012-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.springframework.samples.petclinic.e2e.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

/**
 * Page Object for the Owner form (create/update) page
 */
public class OwnerFormPage extends BasePage {

	public OwnerFormPage(Page page) {
		super(page);
	}

	public void navigateToNewOwnerForm() {
		navigateTo("/owners/new");
	}

	public Locator getFirstNameInput() {
		return page.locator("input[name='firstName']");
	}

	public Locator getLastNameInput() {
		return page.locator("input[name='lastName']");
	}

	public Locator getAddressInput() {
		return page.locator("input[name='address']");
	}

	public Locator getCityInput() {
		return page.locator("input[name='city']");
	}

	public Locator getTelephoneInput() {
		return page.locator("input[name='telephone']");
	}

	public Locator getSubmitButton() {
		return page.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
				new Page.GetByRoleOptions().setName("Add Owner").setExact(false));
	}

	public Locator getUpdateButton() {
		return page.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
				new Page.GetByRoleOptions().setName("Update Owner").setExact(false));
	}

	public void fillOwnerForm(String firstName, String lastName, String address, String city, String telephone) {
		getFirstNameInput().fill(firstName);
		getLastNameInput().fill(lastName);
		getAddressInput().fill(address);
		getCityInput().fill(city);
		getTelephoneInput().fill(telephone);
	}

	public void submitForm() {
		if (getSubmitButton().isVisible()) {
			getSubmitButton().click();
		}
		else {
			getUpdateButton().click();
		}
	}

	public boolean hasValidationErrors() {
		return page.locator(".has-error, .error, .is-invalid").count() > 0;
	}

	public String getValidationError(String fieldName) {
		return page.locator("input[name='" + fieldName + "']")
			.locator("xpath=following-sibling::span[@class='help-block']")
			.textContent();
	}

	public boolean isEditMode() {
		return getUpdateButton().isVisible();
	}

}

File: src/test/playwright/pages/OwnerFormPage.js
const { BasePage } = require('./BasePage');

/**
 * OwnerFormPage class for creating and updating owners
 */
class OwnerFormPage extends BasePage {
  constructor(page) {
    super(page);
  }

  /**
   * Navigate to new owner form
   */
  async gotoNew() {
    await super.goto('/owners/new');
  }

  /**
   * Navigate to edit owner form
   */
  async gotoEdit(ownerId) {
    await super.goto(`/owners/${ownerId}/edit`);
  }

  /**
   * Fill owner form with data
   */
  async fillOwnerForm(ownerData) {
    if (ownerData.firstName) {
      await this.page.locator('#firstName').fill(ownerData.firstName);
    }
    if (ownerData.lastName) {
      await this.page.locator('#lastName').fill(ownerData.lastName);
    }
    if (ownerData.address) {
      await this.page.locator('#address').fill(ownerData.address);
    }
    if (ownerData.city) {
      await this.page.locator('#city').fill(ownerData.city);
    }
    if (ownerData.telephone) {
      await this.page.locator('#telephone').fill(ownerData.telephone);
    }
  }

  /**
   * Submit the form
   */
  async submitForm() {
    await this.page.locator('button[type="submit"]').click();
  }

  /**
   * Check if form is visible
   */
  async isFormVisible() {
    return await this.page.locator('#add-owner-form').isVisible();
  }

  /**
   * Get form title
   */
  async getFormTitle() {
    return await this.page.locator('h2').textContent();
  }

  /**
   * Check if field has validation error
   */
  async hasFieldError(fieldName) {
    const fieldGroup = this.page.locator(`#${fieldName}`).locator('..').locator('..');
    return await fieldGroup.locator('.has-error').isVisible();
  }

  /**
   * Get validation error for a specific field
   */
  async getFieldError(fieldName) {
    const fieldGroup = this.page.locator(`#${fieldName}`).locator('..').locator('..');
    const errorElement = fieldGroup.locator('.help-inline');
    return await errorElement.textContent();
  }

  /**
   * Clear all form fields
   */
  async clearForm() {
    await this.page.locator('#firstName').fill('');
    await this.page.locator('#lastName').fill('');
    await this.page.locator('#address').fill('');
    await this.page.locator('#city').fill('');
    await this.page.locator('#telephone').fill('');
  }

  /**
   * Check if submit button text matches expected value
   */
  async getSubmitButtonText() {
    return await this.page.locator('button[type="submit"]').textContent();
  }
}

module.exports = { OwnerFormPage };